### PR TITLE
test: Test that low difficulty chain fork is rejected

### PR DESCRIPTION
**Pull request by maflcko**
To prevent OOM, Bitcoin Core will reject chain forks at low difficulty by default. This is the only use-case of checkpoints, so add a test for it to make sure the feature works as expected. If it didn't work, checkpoints would have no use-case and we might as well remove them

### CODE REVIEW

**Comment by practicalswift** [comment]
Concept ACK

**Comment by laanwj** [comment]
Concept ACK

**Comment by jamesob** [comment]
Concept ACK, will review in next few days.

**Comment by MarcoFalke** [comment]
Moved test in the test list to a different position to solve the merge conflict without a rebase

**Comment by Sjors** [comment]
Concept ACK, but I find the test hard to read atm.  I'm confused about what's actually in `blockheader_testnet3.hex`. Are those the headers without the block hash of the real testnet3 chain?  Because I don't see the checkpoint hash in there. Also, rather than having two lines with `fork:` and a parser to deal with that, maybe just hard code those in the test itself? Alternatively, create a seperate file for the fork? At what height does this fork start? Genesis or same height as checkpoint?

**Comment by MarcoFalke** [comment]
Added an inline comment to hopefully explain it better, @Sjors 

> Are those the headers without the block hash of the real testnet3 chain? Because I don't see the checkpoint hash in there.

Serialized headers include only the hash of the previous block, so if the last header is the checkpoint, the block hash of the checkpoint will never appear in the data file.

File: test/functional/p2p_dos_header_tree.py [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+            'branchlen': 546,
+            'status': 'headers-only',
+        } in self.nodes[0].getchaintips()
+
+        self.log.info("Feed all fork headers (fails)")
+        with self.nodes[0].assert_debug_log(['bad-fork-prior-to-checkpoint (code 67)']):
+            self.nodes[0].p2p.send_message(msg_headers(self.headers_fork))
+            self.nodes[0].p2p.wait_for_disconnect()
+
+        self.log.info("Feed all fork headers (succeeds)")
```
Comment (Sjors) [inline test/functional/p2p_dos_header_tree.py]: Nit: "Disable checkpoints and feed low work fork headers again (succeeds)"

File: test/functional/p2p_dos_header_tree.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+        self.nodes[0].p2p.send_message(msg_headers(self.headers))
+        self.nodes[0].p2p.sync_with_ping()
+        assert {
+            'height': 546,
+            'hash': '000000002a936ca763904c3c35fce2f3556c559c0214345d31b1bcebf76acb70',
+            'branchlen': 546,
+            'status': 'headers-only',
+        } in self.nodes[0].getchaintips()
+
+        self.log.info("Feed all fork headers (fails)")
```
Comment (Sjors) [inline test/functional/p2p_dos_header_tree.py]: Nit: "Feed headers of fork with less PoW (fails due to checkpoint)"

File: test/functional/p2p_dos_header_tree.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+        # The headers data is taken from testnet3 for early blocks from genesis until the first checkpoint. There are
+        # two headers with valid POW at height 1 and 2, forking off from genesis. They are indicated by the FORK_PREFIX.
+        FORK_PREFIX = 'fork:'
+        self.headers = [l for l in h_lines if not l.startswith(FORK_PREFIX)]
+        self.headers_fork = [l for l in h_lines if l.startswith(FORK_PREFIX)]
+
+        self.headers = [FromHex(CBlockHeader(), h) for h in self.headers]
+        self.headers_fork = [FromHex(CBlockHeader(), h) for h in self.headers_fork]
+
+        self.log.info("Feed all non-fork headers")
```
Comment (Sjors) [inline test/functional/p2p_dos_header_tree.py]: Nit: "Feed non-fork headers, up to and including the first checkpoint (succeeds)"

**Review by Sjors — APPROVED** [review]
ACK fa0c247, but added some suggestions for clearer wording.

To make the behavior more clear, it helps to include non DoS examples. Can you generate a testnet fork with *more* proof of work? It should be shorter than the real chain so it doesn't reach checkpoint height.

With and without checkpoints enabled I would expect this fork to be accepted if it's the first thing we see.

If we see it after we've already seen headers up to the checkpoint, then with checkpoints enabled I expect this to be rejected. With checkpoints disabled, I expect a reorg.

**Comment by MarcoFalke** [comment]
> With and without checkpoints enabled I would expect this fork to be accepted if it's the first thing we see.

Added test case (and node 1). Also, fixed up your comment suggestions.

**Comment by Sjors** [comment]
Thanks for adding the node 1 example. Code review ACK 333317c

**Pull request merged**